File: Problems/0100_0199/0110_Balanced_Binary_Tree/Project_Python3/TreeNode/OperateTreeNode.py
from .TreeNode import TreeNode

class OperateTreeNode:
    def createTreeNode(self, flds):
        return self.createSubTreeNode(flds.split(","), 0, 0)

    def createSubTreeNode(self, flds, depth, pos):
        if len(flds) <= 0:
            return None

        cur_pos = 0
        for i in range(depth):
            cur_pos += 2 ** i

        if cur_pos + pos > len(flds) - 1:
            return None

        if flds[cur_pos + pos] == 'null':
            return None

        node = TreeNode(int(flds[cur_pos + pos]))
        node.left = self.createSubTreeNode(flds, depth + 1, 2*pos)
        node.right = self.createSubTreeNode(flds, depth + 1, 2*pos + 1)

        return node

    def treeToStaircaseString(self, node):
        self.resultList = []
        self.getStaircaseSubString(node, 0)
        return self.resultListToString()

    def getStaircaseSubString(self, node, n):
        if node == None:
            return
        if len(self.resultList) <= n:
            self.resultList.append("(" + str(node.val) + ")")
        else:
            self.resultList[n] += ",(" + str(node.val) + ")"
        if node.left != None:
            self.getStaircaseSubString(node.left, n + 1)
        if node.right != None:
            self.getStaircaseSubString(node.right, n + 1)
        return

    def resultListToString(self):
        resultStr = "\n".join(self.resultList)
        self.resultList.clear()
        return resultStr

    def tree2str(self, node):
        """
        :type node: TreeNode
        :rtype: str
        """
        if node == None:
            return ""

        resultStr = str(node.val)

        if node.left == None and node.right == None:
            return resultStr

        resultStr += "(" + self.tree2str(node.left) + ")"
        if node.right != None:
            resultStr += "(" + self.tree2str(node.right) + ")"

        return resultStr

File: Problems/0900_0999/0993_Cousins_in_Binary_Tree/Project_Java/OperateTreeNode.java
import java.util.ArrayList;
import java.util.List;

public class OperateTreeNode {
    public TreeNode createTreeNode(String flds) {
         return createTreeNode(flds.split(","), 0, 0);
    }

    public TreeNode createTreeNode(String[] flds, int depth, int pos) {
        if (flds.length == 0)
            return null;

        int cur_pos = 0;
        for (int i = 0; i < depth; ++i)
            cur_pos += (int)Math.pow(2, i);
        
        if (cur_pos + pos > flds.length - 1)
            return null;
        
        if (flds[cur_pos + pos].equals("null"))
            return null;

        try {
            TreeNode node = new TreeNode(Integer.parseInt(flds[cur_pos + pos]));
            node.left = createTreeNode(flds, depth + 1, 2*pos);
            node.right = createTreeNode(flds, depth + 1, 2*pos + 1);

            return node;

        } catch (Exception e) {
            System.out.println("\n" +  e + "\n" +
                              "createTreeNode() Error ... flds[" + Integer.toString(cur_pos + pos) + "] = " + flds[cur_pos + pos] + "\n");
            System.exit(1);

            return null;

        }
    }

    List<String> resultList;

    public String treeToStaircaseString(TreeNode node) {
        resultList = new ArrayList<String>();

        String resultStr = treeToStaircaseSubString(node, 0);
        resultList.clear();

        return resultStr;
    }

    private String treeToStaircaseSubString(TreeNode node, int n) {
        if (node == null)
            return "";
        
        if (resultList.size() <= n)
            resultList.add("(" + Integer.toString(node.val) + ")");
        else
            resultList.set(n, resultList.get(n) + ",(" + Integer.toString(node.val) + ")");

        if (node.left != null)
            treeToStaircaseSubString(node.left, n + 1);
        if (node.right != null)
            treeToStaircaseSubString(node.right, n + 1);

        return String.join("\n", resultList) + "\n";
    }

    public String tree2str(TreeNode node) {
        if (node == null)
            return "";

        String resultStr = Integer.toString(node.val);

        if (node.left == null && node.right == null)
            return resultStr;

        resultStr += "(" + tree2str(node.left) + ")";
        if (node.right != null)
            resultStr += "(" + tree2str(node.right) + ")";

        return resultStr;
    }
}
